central/complianceoperator/v2/report/manager/watcher/scanwatcher.go

```diff
@@ -0,0 +1,239 @@
+	resultC chan *storage.ComplianceOperatorCheckResultV2
+	stopped *concurrency.Signal
+
+	readyQueue  readyQueue[*ScanWatcherResults]
+	scanResults *ScanWatcherResults
+	totalChecks int
+}
+
+// NewScanWatcher creates a new ScanWatcher
+func NewScanWatcher(ctx context.Context, watcherID string, queue readyQueue[*ScanWatcherResults]) *scanWatcherImpl {
+	watcherCtx, cancel := context.WithCancel(ctx)
+	finishedSignal := concurrency.NewSignal()
+	ret := &scanWatcherImpl{
+		ctx:        watcherCtx,
+		cancel:     cancel,
+		scanC:      make(chan *storage.ComplianceOperatorScanV2, defaultChanelSize),
+		resultC:    make(chan *storage.ComplianceOperatorCheckResultV2, defaultChanelSize),
+		stopped:    &finishedSignal,
+		readyQueue: queue,
+		scanResults: &ScanWatcherResults{
+			Ctx:          ctx,
+			WatcherID:    watcherID,
+			CheckResults: set.NewStringSet(),
+		},
+	}
+	timeout := time.NewTimer(defaultTimeout)
+	go ret.run(timeout.C)
+	return ret
+}
+
+// PushScan queues a Scan to be handled
+func (s *scanWatcherImpl) PushScan(scan *storage.ComplianceOperatorScanV2) error {
+	select {
+	case <-s.ctx.Done():
+		return errors.New("The watcher is stopped")
+	case s.scanC <- scan:
+		return nil
+	}
+}
+
+// PushCheckResult queues a CheckResult to be handled
+func (s *scanWatcherImpl) PushCheckResult(result *storage.ComplianceOperatorCheckResultV2) error {
+	select {
+	case <-s.ctx.Done():
+		return errors.New("The watcher is stopped")
+	case s.resultC <- result:
+		return nil
+	}
+}
+
+// Finished indicates whether the watcher is finished or not
+func (s *scanWatcherImpl) Finished() concurrency.ReadOnlySignal {
+	return s.stopped
+}
+
+// Stop the watcher
+func (s *scanWatcherImpl) Stop() {
+	s.cancel()
+}
+
+func (s *scanWatcherImpl) run(timerC <-chan time.Time) {
+	defer func() {
+		s.stopped.Signal()
+		<-s.stopped.Done()
+	}()
+	for {
+		select {
+		case <-s.ctx.Done():
+			log.Infof("Stopping scan watcher for scan")
+			return
+		case <-timerC:
+			log.Warnf("Timeout waiting for the scan %s to finish", s.scanResults.Scan.GetScanName())
+			s.scanResults.Error = ScanTimeoutError
+			s.readyQueue.Push(s.scanResults)
+			return
+		case scan := <-s.scanC:
+			if err := s.handleScan(scan); err != nil {
+				log.Errorf("Unable to handle scan %s: %v", scan.GetScanName(), err)
+			}
+		case result := <-s.resultC:
+			if err := s.handleResult(result); err != nil {
+				log.Errorf("Unable to handle result %s for scan %s: %v", result.GetCheckName(), result.GetScanName(), err)
+			}
+		}
+		if s.totalChecks != 0 && s.totalChecks == len(s.scanResults.CheckResults) {
+			s.readyQueue.Push(s.scanResults)
+			return
+		}
+	}
+}
+
+func (s *scanWatcherImpl) handleScan(scan *storage.ComplianceOperatorScanV2) error {
+	if checkCountAnnotation, found := scan.GetAnnotations()[checkCountAnnotationKey]; found {
+		var numChecks int
+		var err error
+		if numChecks, err = strconv.Atoi(checkCountAnnotation); err != nil {
+			return err
```

Comment:
nit: Maybe we should wrap this error, so that we know the context of the parser failure?